### DISCUSSION TITLE
Is it possible to use USB from my UPS to attach it to PC via WIFI ?

### DISCUSSION THREAD

**Discussion by panayotovip** [tobozo/ESP32-USB-Soft-Host#38] [General]
I have a UPS that I want to monitor.
Where the UPS is, there is no computer to do this.
The UPS has a USB port.
Is there any way I can use ESP32 (ESP32-WROOM-32D) with this project.
To connect UPS to ESP32 with USB.
To connect the ESP32 to a wireless network.
On a computer that is in the local network to monitor this UPS.

**Comment by tobozo** [comment]
>  The UPS has a USB port.

If that port exposes a USB 1.0 HID device then yes it's possible.

**Reply by panayotovip** [reply]
This is a good news. :)
Thank you for уour answer.
Now I just have to learn how to use the library.
I am at basic level for working with arduino ide.
I made the LED of my ESP32 (ESP32-WROOM-32D) blink :)
I would appreciate it if someone has time to help.
The important thing is that there is a solution to my problem.
Now it remains to somehow come to realization.
Thanks again for your reply.

**Reply by tobozo** [reply]
first you need to know the device descriptor (i.e. using `lsusb -v` in linux shell, or [Get-PnpDevice](https://www.shellhacks.com/windows-lsusb-equivalent-powershell/) in windows powershell)

then you need to make sure this device works with USB 1.0, this can be achieved by flashing the [USB_Test](https://github.com/tobozo/ESP32-USB-Soft-Host/blob/main/examples/USB_Test/USB_Test.ino#L90) example **with debug level set to `debug`**, and check the serial output from the ESP (device descriptor is printed upon detection). 

it is recommended to have several USB-LS HID devices to test with (old mouse, keyboard) until at least one of them is detected, this trial/error sequence will hopefully give you at least one positive result you can use as a reference to test if your UPS is indeed a USB-LS device.

if the UPS gets successfully detected, you can then start testing its behaviour (e.g. capture UPS key events as they occur).

**Reply by panayotovip** [reply]
Thank you a lot for starting point and direction.
I went to the UPS and hooked up to it with a cable to get the lsusb -v information for it.
```
Bus 003 Device 003: ID 051d:0002 American Power Conversion Uninterruptible Power Supply
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x051d American Power Conversion
  idProduct          0x0002 Uninterruptible Power Supply
  bcdDevice            1.06
  iManufacturer           3 American Power Conversion
  iProduct                1 Back-UPS XS 950U   FW:925.T1 .I USB FW:T1 
  iSerial                 2 xxxxxxxxx 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0022
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xe0
      Self Powered
      Remote Wakeup
    MaxPower               24mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.10
          bCountryCode           33 US
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength    1028
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0006  1x 6 bytes
        bInterval              10
Device Status:     0x0000
  (Bus Powered)
```
These days I will try to hook ESP32-WROOM-32D to UPS.
I looked at the test file it said to load into the ESP.
For the physical connection I have to use pins
   #define DP_P0 16 // always enabled
   #define DM_P0 17 // always enabled
on the board these pins are labeled as
   RX2
   TX2
I know that GND should also be connected.
Is there a need to connect the 5v from the ESP to the USB on the UPS?

**Reply by tobozo** [reply]
bcdUSB               1.10

looks like it's detected as USB 1.1 (USB-FS), while this format isn't supported by USB-Soft-Host library, there's still a chance it's only the *negotiated* speed between the UPS controller and the linux host controller.

>  Is there a need to connect the 5v from the ESP to the USB on the UPS?

not with the UPS since it has its own power, however may need to connect the 5v pin when testing a joypad, a mouse or a keyboard (provided they don't work on batteries)

**Reply by chegewara** [reply]
actually you cant tell from bcdUSB if its FS or LS.
LS can be compatible with USB 1.1 too.

my suggestion is to use `usbview` tool on linux
